File: SDKs/ios/Sources/KISR/KISRWalletAdapterJS.swift
// ============================================================================
// EXAMPLE FILE
// ----------------------------------------------------------------------------
// Used as an example of how to bridge to your SDK/WASM
// ============================================================================
//
//
import Foundation

public final class KISRWalletAdapterJS: KISRWalletAdapter {
	private let bridge: KISRJSCoreBridge // use as an example of how to bridge to your SDK/WASM
	private let networkId: UInt8
	private let privateKey: String

	public init?(networkId: UInt8 = 0, privateKey: String) {
		guard let bridge = KISRJSCoreBridge() else { return nil }
		self.bridge = bridge
		self.networkId = networkId
		self.privateKey = privateKey
	}

	public func currentNetworkId() -> UInt8 { networkId }

	public func createSelfUtxo(amountSompi: UInt64) throws -> (txid: String, vout: UInt32, amountSompi: UInt64) {
		let network = (networkId == 0) ? "mainnet" : "testnet-10"
		let resAny = try bridge.callAsync("createUtxoToSelf", args: [[
			"privateKey": privateKey,
			"amountSompi": String(amountSompi),
			"network": network
		]])
		guard let res = resAny as? [String: Any], let txid = res["txid"] as? String else { throw KISRError.payloadInvalid }
		let index = (res["index"] as? Int) ?? 0
		let vout = UInt32(index)
		return (txid, vout, amountSompi)
	}

	public func preSignInput(txid: String, vout: UInt32, sighashFlags: UInt8) throws -> Data {
		let network = (networkId == 0) ? "mainnet" : "testnet-10"
		let resAny = try bridge.callAsync("preSignCreatedUtxo", args: [[
			"privateKey": privateKey,
			"network": network,
			"txid": txid,
			"index": Int(vout)
		]])
		guard let res = resAny as? [String: Any], let sigHex = res["signatureHex"] as? String else { throw KISRError.payloadInvalid }
		return Data(hexString: sigHex)
	}

	public func publishAnchorTransaction(payload: Data) throws -> String {
		let network = (networkId == 0) ? "mainnet" : "testnet-10"
		let resAny = try bridge.callAsync("createAnchorToSelfWithPayload", args: [[
			"privateKey": privateKey,
			"network": network,
			"payloadHex": payload.hexString
		]])
		guard let res = resAny as? [String: Any], let txid = res["txid"] as? String else { throw KISRError.payloadInvalid }
		return txid
	}

	public func assembleRedemption(toAddress: String, inputTxid: String, inputIndex: UInt32, inputAmountSompi: UInt64, presigHex: String, feeSompi: UInt64) throws -> Data {
		// Example adapter responsibility: bridge to your SDK/WASM to construct the raw tx.
		// For JS bridge, you would expose a `assembleRedemption` in kisr_core.js and call it here.
		// Returning error to indicate this must be implemented per-wallet.
		throw KISRError.wasmUnavailable("kaspa assembleRedemption not bridged in JS core")
	}

	public func broadcastRedemption(rawTx: Data) throws -> String {
		let network = (networkId == 0) ? "mainnet" : "testnet-10"
		let txid = try bridge.callAsync("submitRawTransactionHex", args: [rawTx.hexString, network])
		guard let id = txid as? String, !id.isEmpty else { throw KISRError.payloadInvalid }
		return id
	}
}

File: SDKs/android/kisr/src/main/java/org/kisr/sdk/KISRWalletAdapterJS.kt
// ============================================================================
// EXAMPLE FILE
// ----------------------------------------------------------------------------
// Used as an example of how to bridge to your SDK/WASM
// ============================================================================
//
//
package org.kisr.sdk

class KISRWalletAdapterJS(
	private val bridge: KISRJSCoreBridge,
	private val networkId: UByte = 0u,
	private val privateKey: String
) : KISRWalletAdapter {
	override fun currentNetworkId(): UByte = networkId

	override fun createSelfUtxo(amountSompi: Long): Triple<String, UInt, Long> {
		val network = if (networkId == 0u.toUByte()) "mainnet" else "testnet-10"
		val res = bridge.call(
			name = "createUtxoToSelf",
			args = listOf(mapOf(
				"privateKey" to privateKey,
				"amountSompi" to amountSompi.toString(),
				"network" to network
			))
		) as? Map<*, *> ?: throw KISRError.PayloadInvalid()
		val txid = (res["txid"] as? String) ?: throw KISRError.PayloadInvalid()
		val index = (res["index"] as? Number)?.toInt() ?: 0
		return Triple(txid, index.toUInt(), amountSompi)
	}

	override fun preSignInput(txid: String, vout: UInt, sighashFlags: UByte): ByteArray {
		val network = if (networkId == 0u.toUByte()) "mainnet" else "testnet-10"
		val res = bridge.call(
			name = "preSignCreatedUtxo",
			args = listOf(mapOf(
				"privateKey" to privateKey,
				"network" to network,
				"txid" to txid,
				"index" to vout.toInt()
			))
		) as? Map<*, *> ?: throw KISRError.PayloadInvalid()
		val sigHex = (res["signatureHex"] as? String) ?: throw KISRError.PayloadInvalid()
		return org.kisr.sdk.internal.hexStringToBytes(sigHex)
	}

	override fun publishAnchorTransaction(payload: ByteArray): String {
		val network = if (networkId == 0u.toUByte()) "mainnet" else "testnet-10"
		val res = bridge.call(
			name = "createAnchorToSelfWithPayload",
			args = listOf(mapOf(
				"privateKey" to privateKey,
				"network" to network,
				"payloadHex" to org.kisr.sdk.internal.bytesToHex(payload)
			))
		) as? Map<*, *> ?: throw KISRError.PayloadInvalid()
		return (res["txid"] as? String) ?: throw KISRError.PayloadInvalid()
	}

	override fun assembleRedemption(
		toAddress: String,
		inputTxid: String,
		inputIndex: UInt,
		inputAmountSompi: Long,
		presigHex: String,
		feeSompi: Long
	): ByteArray {
		throw KISRError.WasmUnavailable("kaspa assembleRedemption not bridged in JS core")
	}

	override fun broadcastRedemption(rawTx: ByteArray): String {
		val network = if (networkId == 0u.toUByte()) "mainnet" else "testnet-10"
		val id = bridge.call(
			name = "submitRawTransactionHex",
			args = listOf(org.kisr.sdk.internal.bytesToHex(rawTx), network)
		) as? String ?: throw KISRError.PayloadInvalid()
		return id
	}
}
